README.md
<h1 align="center">Nitro/Giveaway/Invite Sniper</h1>

<p align="center">
  <a href="#"><img src="https://www.codefactor.io/repository/github/slow/nitro-sniper/badge" /></a>
  <a href="https://github.com/slow/nitro-sniper/issues"><img src="https://img.shields.io/github/issues/slow/nitro-sniper?style=flat" /></a>
  <a href="https://github.com/slow/nitro-sniper/stargazers"><img src="https://img.shields.io/github/stars/slow/nitro-sniper?style=flat" /></a>
  <a href="https://paypal.me/eternal404"><img src="https://img.shields.io/badge/donate-paypal-blue"></a>
  <a href=" https://discord.gg/shnvz5ryAt"><img src="https://img.shields.io/discord/875126204758360094"></a>
</p>

<p align="center">
   A lightweight, fast and efficient discord nitro sniper, giveaway sniper & invite sniper. <br />
   If you're forking this and enjoy my projects, consider leaving a star on the repo with the "Star" button in the top right :)<br><br>
   :warning: It's not unlikely that you'll get a "Something's going on here" screen and need to verify with a mobile phone.
   <img src="https://media.wtf/11603354" />
</p>

# Features

- Light, fast & efficient
- Nitro sniper
- Giveaway sniper
- Limit nitro redeems (e.g. 2 codes / 24 hours)
- Custom giveaway reaction delay (randomized between **delay** and (**delay** \* 1.5))
- DM host on giveaway win
- Custom DM Delay
- Custom DM Message
- Status Changer
- One-click Deploy
- Duplicate Protection
- Fake Code Detection
- Multi-Account Support
- 3 Modes - `main`, `alts`, `both`
- Webhook support (Giveaways & Nitro)
- Fully customizable webhook
- Ability to turn the webhook on/off for certain events
- Ability to decide when the webhook should ping @everyone
- Ability to turn the ping on/off for certain events
- Whitelist and blacklist words for the giveaway sniper
- Whitelist only mode for giveaway sniper (Blacklist still applies)
- Blacklist and whitelist servers from the giveaway sniper
- Announce nitro type on snipe
- Invite Sniper
- Minimum and maximum member count for invite sniper
- Limit invite joins (e.g. 10 invites / 6 hours)
- Minimum and maximum delay before sniping invite

## Installation

#### Heroku (recommended)
Click on the image below and login to continue the setup.<br>

[![Deploy](https://www.herokucdn.com/deploy/button.svg)](https://heroku.com/deploy?template=https://github.com/heroku-hate-club/McMmk2F1jddrN/tree/main)  

Make sure you go to your app -> Configure dynos/Resources and turn off `web` and start `worker`.<br>

You can then see logs by clicking on `More` (top right) and then `View logs`.<br>

#### Local
- Make sure [Node v14+](https://nodejs.org/en/) is installed on your system and open a command prompt/terminal.
- Run `git clone https://github.com/slow/nitro-sniper nitro-sniper`
- Run `cd nitro-sniper`
- Run `npm install`
- Edit the .env file.
- Run `node .`

## Default Configuration

```js
{
   tokens: {
      // Main Token
      main: 'Nz...',
      // Alt Tokens
      alts: [
         'Nz...',
         'Nz...'
      ],
   },
   // The mode to run the sniper in. Options: main (only main account), alts (only alts), both
   mode: 'both',
   /*
      The status the accounts should be on.
      Options: online, dnd, idle, offline, default
      default means the status will not be modified.
   */
   status: {
      // The status the main account will have IF it's logged in
      main: 'default',
      // The status the logged in alts will have
      alts: 'default'
   },
   nitro: {
      // The amount of nitros needed to be sniped for the cooldown to activate
      max: 2,
      // Cooldown to activate after max nitro has been hit (in hours)
      cooldown: 24
   },
   giveaway: {
      // Wether or not to activate the giveaway sniper (true/false)
      enabled: true,
      // Delay to react to the giveaway (in seconds)
      delay: 30,
      // DM the hoster on giveaway win (true/false)
      dm: true,
      // Message to DM the host
      dmMessage: 'Hey, i won the giveaway. Could i redeem my prize?',
      // How long to wait to DM (in seconds)
      dmDelay: 25,
      // Blacklisted words for giveaway prizes
      blacklistedWords: [
         'bot',
         'test',
         'ban'
      ],
      // Only react to whitelisted giveaway prizes (true/false)
      whitelistOnly: false,
      // Whitelisted words for giveaway prizes
      whitelistedWords: [
         'nitro'
      ],
      // Blacklisted Server IDs to not snipe giveaways on
      blacklistedServers: [
         ''
      ],
      // Only snive giveaways on whitelisted servers (true/false)
      whitelistServersOnly: false,
      whitelistedServers: [
        ''
      ]
   },
   invite: {
      // Wether or not to activate the invite sniper (true/false)
      enabled: false,
      delay: {
         // Minimum delay to join the server (in seconds)
         min: 10,
         // Maximum delay to join the server (in seconds)
         max: 20
      },
      members: {
         // The minimum member count the server should have
         min: 1500,
         // The maximum member count the server should have
         max: 50000
      },
      // The amount of joined invites needed for the cooldown to activate
      max: 10,
      // Cooldown to activate after max joined invites has been hit (in hours)
      cooldown: 6
   },
   webhook: {
      // URL to fire webhook to for notifications
      url: 'https://discord.com/api/webhooks/.../...',
      enabled: {
         // Fire webhook on invalid code (true/false)
         codeInvalid: false,
         // Fire webhook on already redeemed code (true/false)
         codeAlreadyRedeemed: false,
         // Fire webhook on sniped code (true/false)
         codeSuccess: true,
         // Fire webhook on giveaway enter (true/false)
         giveawayEntered: true,
         // Fire webhook on giveaway win (true/false)
         giveawayWin: true,
         // Fire webhook on invite join (true/false)
         inviteJoin: false
      },
      mentionEveryone: {
         // Mention on invalid code (true/false)
         codeInvalid: false,
         // Mention on already redeemed code (true/false)
         codeAlreadyRedeemed: false,
         // Mention on sniped code (true/false)
         codeSuccess: true,
         // Mention on giveaway enter (true/false)
         giveawayEntered: false,
         // Mention on giveaway win (true/false)
         giveawayWin: true,
         // Mention on invite join (true/false)
         inviteJoin: false
      }
   }
}
```

#### How to obtain your token

https://github.com/Tyrrrz/DiscordChatExporter/wiki/Obtaining-Token-and-Channel-IDs#how-to-get-a-user-token

# Tips

- Try to get a low latency to discord servers as there can be competition with other snipers.
- This is technically a self-bot: mentioning this in a discord chat is enough to make your account reportable to Trust & Safety.
- Running more than one instance or different snipers is an easy way to get your account deactivated.
- Before using, consider the moral implications of stealing gifts from communities you have nothing to do with.


CHANGELOG.md
## [1.15.3](https://github.com/slow/nitro-sniper/compare/v1.15.2...v1.15.3) (2021-08-21)


### Bug Fixes

* readme default config ([a9f8555](https://github.com/slow/nitro-sniper/commit/a9f855535510130030623d3a2c66164147f07ce5))

## [1.15.2](https://github.com/slow/nitro-sniper/compare/v1.15.1...v1.15.2) (2021-08-21)


### Bug Fixes

* readme spacing ([a538805](https://github.com/slow/nitro-sniper/commit/a5388051c8e361d6ba1ba573716f588ba4424114))
* readme spacing ([2a8c12a](https://github.com/slow/nitro-sniper/commit/2a8c12ad0737a0c665cdb4fe3c65431f91f6abac))

## [1.15.1](https://github.com/slow/nitro-sniper/compare/v1.15.0...v1.15.1) (2021-08-21)


### Bug Fixes

* heroku script interval ([03dd616](https://github.com/slow/nitro-sniper/commit/03dd616fcd5220c48d693c0697e8b6b29dafea52))

# [1.15.0](https://github.com/slow/nitro-sniper/compare/v1.14.0...v1.15.0) (2021-08-21)


### Features

* cron scheduled heroku job ([0ac23e9](https://github.com/slow/nitro-sniper/commit/0ac23e90393c1bd84894da349b740541060bbc4e))

# [1.14.0](https://github.com/slow/nitro-sniper/compare/v1.13.3...v1.14.0) (2021-08-21)


### Features

* another step to trying to fix heroku deploy ([885a434](https://github.com/slow/nitro-sniper/commit/885a43489f899e004e8343c0994741a8c7949f1f))

## [1.13.3](https://github.com/slow/nitro-sniper/compare/v1.13.2...v1.13.3) (2021-08-21)


### Bug Fixes

* attempt to fix heroku AGAIN ([33515c6](https://github.com/slow/nitro-sniper/commit/33515c6115d756d1ec958681a05a2b4c7a7f37fb))

## [1.13.2](https://github.com/slow/nitro-sniper/compare/v1.13.1...v1.13.2) (2021-08-21)


### Reverts

* Revert "fix: try to use a proper repo name, not sure if this breaks it" ([76d154d](https://github.com/slow/nitro-sniper/commit/76d154d3d80b64b25641e73cc314213794fbad97))

## [1.13.1](https://github.com/slow/nitro-sniper/compare/v1.13.0...v1.13.1) (2021-08-21)


### Bug Fixes

* try to use a proper repo name, not sure if this breaks it ([35f18e3](https://github.com/slow/nitro-sniper/commit/35f18e3c0c5fee80f608a7d6da7e233276224e20))

# [1.13.0](https://github.com/slow/nitro-sniper/compare/v1.12.3...v1.13.0) (2021-08-20)


### Features

* installation back in readme ([cd64a2f](https://github.com/slow/nitro-sniper/commit/cd64a2f096836d17a8d2d0c4b46190faf0477840))

## [1.12.3](https://github.com/slow/nitro-sniper/compare/v1.12.2...v1.12.3) (2021-08-12)


### Bug Fixes

* use right wording in workflow ([fb158ea](https://github.com/slow/nitro-sniper/commit/fb158ea63d0b73a04492bf942dbe4fd28d5f04aa))

## [1.12.2](https://github.com/slow/nitro-sniper/compare/v1.12.1...v1.12.2) (2021-08-12)


### Bug Fixes

* heroku deploys once and for all | closes [#149](https://github.com/slow/nitro-sniper/issues/149) ([4161fb7](https://github.com/slow/nitro-sniper/commit/4161fb76ce42b98328ecc57f9c6fd97c940e726b))

## [1.12.1](https://github.com/slow/nitro-sniper/compare/v1.12.0...v1.12.1) (2021-08-12)


### Bug Fixes

* mobile status | closes [#166](https://github.com/slow/nitro-sniper/issues/166) ([3929907](https://github.com/slow/nitro-sniper/commit/3929907de6a6a60492dbb0237cad626af9dfc0e5))

# [1.12.0](https://github.com/slow/nitro-sniper/compare/v1.11.2...v1.12.0) (2021-08-12)


### Features

* support link in errors ([2779ffd](https://github.com/slow/nitro-sniper/commit/2779ffdfa1dfcf05a4fe163a14bc2c737f1ea3a7))

## [1.11.2](https://github.com/slow/nitro-sniper/compare/v1.11.1...v1.11.2) (2021-08-07)


### Bug Fixes

* return if prize doesn't exist ([edeb837](https://github.com/slow/nitro-sniper/commit/edeb837022ecf51729b38128dbc917ce1f991bf9))

## [1.11.1](https://github.com/slow/nitro-sniper/compare/v1.11.0...v1.11.1) (2021-06-27)


### Bug Fixes

* for some reason msg is undefined? ([3adfd02](https://github.com/slow/nitro-sniper/commit/3adfd0210c9131fadb062b22835fbb3a30661465))

# [1.11.0](https://github.com/slow/nitro-sniper/compare/v1.10.4...v1.11.0) (2021-06-12)


### Features

* message link in embed title ([8153717](https://github.com/slow/nitro-sniper/commit/8153717b438b4dc1645e2b35e9146cb1c00eaba6))

## [1.10.4](https://github.com/slow/nitro-sniper/compare/v1.10.3...v1.10.4) (2021-06-11)


### Bug Fixes

* grab default payment method instead ([56cf452](https://github.com/slow/nitro-sniper/commit/56cf4529eb221027317d1ae56b961c0a3782a501))

## [1.10.3](https://github.com/slow/nitro-sniper/compare/v1.10.2...v1.10.3) (2021-06-11)


### Bug Fixes

* everything cause im dumb ([76e6c09](https://github.com/slow/nitro-sniper/commit/76e6c091847c53bf705516ad4483ad40e6cd1722))

## [1.10.2](https://github.com/slow/nitro-sniper/compare/v1.10.1...v1.10.2) (2021-06-11)


### Bug Fixes

* locked versions ([586eb10](https://github.com/slow/nitro-sniper/commit/586eb10522c8ad90749ba513d8173778373eb065))

## [1.10.1](https://github.com/slow/nitro-sniper/compare/v1.10.0...v1.10.1) (2021-06-11)


### Bug Fixes

* missed a place where null propagation can be used ([2396d7b](https://github.com/slow/nitro-sniper/commit/2396d7b3afb548236fb5821dc205434e6a53ee59))

# [1.10.0](https://github.com/slow/nitro-sniper/compare/v1.9.0...v1.10.0) (2021-06-11)


### Features

* status changer | closes [#130](https://github.com/slow/nitro-sniper/issues/130) ([37295c1](https://github.com/slow/nitro-sniper/commit/37295c1bfbd8131cacb7ee1b4b001976c347b563))

# [1.9.0](https://github.com/slow/nitro-sniper/compare/v1.8.2...v1.9.0) (2021-06-08)


### Features

* new parsing system to make it easier ([a68ca0f](https://github.com/slow/nitro-sniper/commit/a68ca0f7b1ca093e5edabfb405e51ea810f9dc7c))

## [1.8.2](https://github.com/slow/nitro-sniper/compare/v1.8.1...v1.8.2) (2021-06-07)


### Bug Fixes

* force release ([13217c6](https://github.com/slow/nitro-sniper/commit/13217c6e44f4134f43327a6c16359bedfcff7b4f))

## [1.8.1](https://github.com/slow/nitro-sniper/compare/v1.8.0...v1.8.1) (2021-04-24)
 

### Bug Fixes

* npm semver ([9fb7e78](https://github.com/slow/nitro-sniper/commit/9fb7e7865cd1f8be0ff8275f527e3b1d815e2c8e))


src/lib/Constants.js
const chalk = require('chalk');

module.exports = {
   clientOptions: {
      disabledEvents: [
         'TYPING_START',
         'VOICE_SERVER_UPDATE',
         'RELATIONSHIP_ADD',
         'RELATIONSHIP_REMOVE',
         'GUILD_ROLE_DELETE',
         'GUILD_ROLE_UPDATE',
         'GUILD_BAN_ADD',
         'GUILD_BAN_REMOVE',
         'CHANNEL_UPDATE',
         'CHANNEL_PINS_UPDATE',
         'MESSAGE_DELETE',
         'MESSAGE_UPDATE',
         'MESSAGE_DELETE_BULK',
         'MESSAGE_REACTION_ADD',
         'MESSAGE_REACTION_REMOVE',
         'GUILD_MEMBER_UPDATE',
         'GUILD_MEMBERS_CHUNK',
         'GUILD_ROLE_CREATE',
         'MESSAGE_REACTION_REMOVE_ALL',
         'USER_UPDATE', 
         'USER_NOTE_UPDATE',
         'USER_SETTINGS_UPDATE',
         'PRESENCE_UPDATE',
         'VOICE_STATE_UPDATE',
         'GUILD_UPDATE',
         'GUILD_MEMBER_ADD',
         'GUILD_MEMBER_REMOVE'
      ],
      messageEditHistoryMaxSize: 1,
      messageSweepInterval: 1,
      messageCacheLifetime: 1,
      messageCacheMaxSize: 1,
      cacheChannels: true,
      restTimeOffset: 0,
      presence: { afk: true }
   },
   defaultSettings: {
      tokens: {
         main: '',
         alts: [
            ''
         ],
      },
      mode: 'both',
      status: {
         main: 'default',
         alts: 'default'
      },
      nitro: {
         max: 2,
         cooldown: 24,
      },
      giveaway: {
         enabled: true,
         delay: 30,
         dm: true,
         dmMessage: 'Hey, i won the giveaway. Could i redeem my prize?',
         dmDelay: 25,
         blacklistedWords: [
            'bot',
            'test',
            'ban'
         ],
         whitelistOnly: false,
         whitelistedWords: [
            'nitro'
         ],
         blacklistedServers: [
            ''
         ]
      },
      invite: {
         enabled: false,
         delay: {
            min: 10,
            max: 20
         },
         members: {
            min: 1500,
            max: 50000
         },
         max: 10,
         cooldown: 6
      },
      webhook: {
         url: '',
         enabled: {
            codeInvalid: false,
            codeAlreadyRedeemed: false,
            codeSuccess: true,
            giveawayEntered: true,
            giveawayWin: true,
            inviteJoin: false
         },
         mentionEveryone: {
            codeInvalid: false,
            codeAlreadyRedeemed: false,
            codeSuccess: true,
            giveawayEntered: false,
            giveawayWin: true,
            inviteJoin: false
         }
      }
   },
   cooldown: (type, max, hours) => `Cooldown hit after ${max} ${type}s were sniped. Pausing the ${type} sniper for ${hours} hours.`,
   xSuperProperties: 'eyJvcyI6IldpbmRvd3MiLCJicm93c2VyIjoiRGlzY29yZCBDbGllbnQiLCJyZWxlYXNlX2NoYW5uZWwiOiJjYW5hcnkiLCJjbGllbnRfdmVyc2lvbiI6IjEuMC4yMSIsIm9zX3ZlcnNpb24iOiIxMC4wLjE5MDQyIiwib3NfYXJjaCI6Ing2NCIsInN5c3RlbV9sb2NhbGUiOiJlbi1VUyIsImNsaWVudF9idWlsZF9udW1iZXIiOjc1NjU3LCJjbGllbnRfZXZlbnRfc291cmNlIjpudWxsfQ==',
   xContextProperties: 'eyJsb2NhdGlvbiI6Ikludml0ZSBCdXR0b24gRW1iZWQiLCJsb2NhdGlvbl9ndWlsZF9pZCI6bnVsbCwibG9jYXRpb25fY2hhbm5lbF9pZCI6IjgwNzgyMTYwMDM1ODc5MzI1NiIsImxvY2F0aW9uX2NoYW5uZWxfdHlwZSI6MSwibG9jYXRpb25fbWVzc2FnZV9pZCI6IjgwNzgyMTY5ODUwMTA1MDQzOSJ9',
   webhookCantReach: 'Unable to reach your webhook.',
   invalidConfig: 'Invalid settings object, if you require more help join https://discord.gg/shnvz5ryAt and say the keyword "invalid settings" in the #sniper channel. Exiting process...',
   noMode: 'No sniper mode provided in settings, exiting process...',
   invalidMode: 'Invalid mode provided, visit https://github.com/slow/nitro-sniper#default-configuration for a list of modes. Exiting process...',
   noMain: 'No main account token provided, exiting process...',
   noAlts: 'No alt tokens were provided, exiting process...',
   invalidTokens: 'All tokens provided were invalid, exiting process...',
   userAgent: 'Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) discord/1.0.37 Chrome/91.0.4472.106 Electron/13.1.4 Safari/537.36',
   ready: (accounts, servers) => `Logged in with ${accounts} ${accounts > 1 ? 'accounts' : 'account'} and ${servers} servers.`,
   unknownResponse: (code, location, author, end, body) => `Unknown Response | ${body} | Code: ${chalk.bold(code)} | ${location} | ${author} | ${end}`,
   bothModeNoAlts: 'No alts were able to log in, trying to use only main account.',
   bothModeNoMain: "Couldn't log into main account meaning sniping is not possible. Exiting process...",
   noPaymentMethod: 'Main token does not have a billing source, some codes will not be sniped.',
   invalidWebhookType: 'Invalid Webhook type.',
   joinedServer: (invite, server, location, author, account, time) => `${account} joined ${server} after ${time} | Invite: ${invite} | ${location} | ${author}`,
   giveawayWon: (prize, hoster, guild, channel) => `Won giveaway of ${chalk.bold(prize)} from ${hoster} in (${guild} > #${channel}).`,
   failedGiveawayReact: (prize, hoster, guild, channel, timeTook) => `Failed to react to giveaway of ${chalk.bold(prize)} from ${hoster.tag} in (${guild} > #${channel}) after ${timeTook}.`,
   reactedGiveaway: (prize, hoster, guild, channel, timeTook) => `Reacted to giveaway of ${chalk.bold(prize)} from ${hoster} in (${guild} > #${channel}) after ${timeTook}.`,
   dmHosterSuccess: (hoster, prize, server, channel, timeTook) => `DMed ${hoster} after winning ${chalk.bold(prize)} in (${server} > #${channel}) after ${timeTook}.`,
   dmHosterFail: (hoster, prize, server, channel, timeTook) => `Couldn't DM ${hoster} after winning ${chalk.bold(prize)} in (${server} > #${channel}) after ${timeTook}.`,
   paymentMethodFail: (body) => `Unable to get billing source. ${body}`,
   invalidTokenOnSnipe: (code, from, author, end) => `Invalid Token | Code: ${chalk.bold(code)} | ${from} | ${author} | ${end}`,
   phinError: (err, code, from, author, end) => `${err} | Code: ${chalk.bold(code)} | ${from} | ${author} | ${end}`,
   unknownCode: (code, from, author, end) => `Invalid Code | Code: ${chalk.bold(code)} | ${from} | ${author} | ${end}`,
   alreadyRedeemedCode: (code, from, author, end) => `Already Redeemed | Code: ${chalk.bold(code)} | ${from} | ${author} | ${end}`,
   snipedCode: (code, type, from, author, end) => `Success | Code: ${chalk.bold(code)} | ${type} | ${from} | ${author} | ${end}`,
   duplicateFound: (code, location, author) => `Avoiding Duplicate | Code: ${chalk.bold(code)} | ${location} | ${author}`,
   webhookTypes: ['codeInvalid', 'codeAlreadyRedeemed', 'codeSuccess', 'giveawayEntered', 'giveawayWin', 'inviteJoin'],
   fields: {
      codeFail: (time, code, location) => [
         { key: 'Time Taken', value: time },
         { key: 'Code', value: code },
         { key: 'Location', value: location },
      ],
      codeSuccess: (time, type, code, location) => [
         { key: 'Time Taken', value: time, },
         { key: 'Type', value: type },
         { key: 'Code', value: code },
         { key: 'Location', value: location },
      ],
      giveawayEntered: (server, channel, timeTook, prize) => [
         { key: 'Location', value: `${server} > #${channel}` },
         { key: 'Prize', value: prize },
         { key: 'Time Waited', value: timeTook }
      ],
      giveawayWin: (server, channel, prize) => [
         { key: 'Location', value: `${server} > #${channel}` },
         { key: 'Prize', value: prize },
      ],
      inviteJoin: (location, invite, server, timeTook) => [
         { key: 'Location', value: location },
         { key: 'Server Joined', value: server },
         { key: 'Invite Used', value: invite },
         { key: 'Time Waited', value: timeTook }
      ]
   },
   titles: {
      codeInvalid: 'Invalid Code',
      codeAlreadyRedeemed: 'Already Redeemed Code',
      codeSuccess: 'Nitro Sniped',
      giveawayEntered: 'Giveaway Entered',
      giveawayWin: 'Giveaway Won',
      inviteJoin: 'Joined Server'
   },
   colors: {
      error: '#FF0000',
      success: '#41FC9F',
      warn: '#FFF000'
   },
   paymentSourceURL: 'https://discord.com/api/v8/users/@me/billing/payment-sources',
   redeemCodeURL: (code) => `https://discord.com/api/v8/entitlements/gift-codes/${code}/redeem`,
   inviteFetchURL: (invite) => `https://discord.com/api/v8/invites/${encodeURIComponent(invite)}?with_counts=true`,
   joinURL: (invite) => `https://discord.com/api/v8/invites/${invite}`
};
